### ISSUE TITLE
Uncaught TypeError: Cannot read property 'MeteorX' of undefined

### ISSUE THREAD

**Issue by peterkuiper** [meteorhacks/meteorx#5]
For some reason, meteorx v1.4.1 is causing problems for us:

Uncaught TypeError: Cannot read property 'MeteorX' of undefined
Uncaught TypeError: Cannot read property 'ServiceConfiguration' of undefined
Uncaught ReferenceError: Spacebars is not defined
Uncaught ReferenceError: Spacebars is not defined

Disabling meteorhacks:kadira in .meteor/packages and manually changing the version in .meteor/versions for meteorhacks:meteorx to 1.0.2 seems to fix this problem.

We only noticed this problem a few days ago because running normal 'meteor' during development works fine. Only when you create a bundle and deploy or run 'meteor --production' does this cause problems.

It seems meteorhacks:kadira really needs meteorhacks:meteorx 1.4.1 because I can only change the version after I disable meteorhacks:kadira in .meteor/packages. If I try to change .meteor/versions to 1.0.2 the meteor package manager automagically upgrades to 1.4.1 (again).

We are running meteor version 1.1.0.3 for this project, 

If you need more info let me know.

**Comment by peterkuiper** [comment]
It seems the problem is caused by Package.onUse and Package.onTest. I have added the repo into the packages dir and applied the following changes:

```
diff --git a/package.js b/package.js
index 506537d..21249e5 100644
--- a/package.js
+++ b/package.js
@@ -5,12 +5,12 @@ Package.describe({
   "name": "meteorhacks:meteorx"
 });

-Package.onUse(function(api) {
+Package.on_use(function(api) {
   configurePackage(api);
   api.export(['MeteorX']);
 });

-Package.onTest(function(api) {
+Package.on_test(function(api) {
   configurePackage(api);
   api.use([
     'tinytest',
```

This has fixed the errors so I assume it's related to our version of meteor for this project. It seems 1.1.0.3 doesn't recognize the newer syntax of onUse and/or onTest. Updating to 1.2.x also fixes the problem so it seems likely 1.1.x is at fault.

**Comment by peterkuiper** [comment]
After further inspection it seems that the issue is unrelated to on_use/onUse. Simple cloning your repo into the project's packages directory (@v1.4.1) fixes the problem which makes me believe that it could be some load order issue. As I'm relatively new to Meteor I'm not up-to-speed on the internals so I'm just guessing at this point. It seems the packages dir has precendence over non local packages.

Here is a diff between build1 (working) and the non working build:

```
diff -r ../build1 .meteor/local/build
diff -r ../build1/programs/server/packages/meteorhacks_meteorx.js .meteor/local/build/programs/server/packages/meteorhacks_meteorx.js
12c12
< (function () {
---
> (function(){
16c16
< // packages/meteorhacks:meteorx/lib/livedata.js                                                 //
---
> // packages/meteorhacks_meteorx/lib/livedata.js                                                 //
82c82
< (function () {
---
> (function(){
86c86
< // packages/meteorhacks:meteorx/lib/mongo-livedata.js                                           //
---
> // packages/meteorhacks_meteorx/lib/mongo-livedata.js                                           //
96c96
<   namespace.MongoConnection = MongoInternals.defaultRemoteCollectionDriver().mongo.constructor; // 7
---
>   namespace.MongoConnection = MongoInternals.defaultRemoteCollectionDriver().mongo.constructor;
150c150
< (function () {
---
> (function(){
154c154
< // packages/meteorhacks:meteorx/lib/server.js                                                   //
---
> // packages/meteorhacks_meteorx/lib/server.js                                                   //
208,209d207
<
< //# sourceMappingURL=meteorhacks_meteorx.js.map
Only in ../build1/programs/server/packages: meteorhacks_meteorx.js.map
diff -r ../build1/programs/server/program.json .meteor/local/build/programs/server/program.json
496,497d495
<       "sourceMap": "packages/meteorhacks_meteorx.js.map",
<       "sourceMapRoot": "packages",
Only in .meteor/local/build/programs/web.browser: 2de97dc80987b5a3279c0d84d23ebcc4b01b3e66.js
Only in ../build1/programs/web.browser: 4a77bd24e01b01378d1cd2099e98ed5ed328ff6f.js
diff -r ../build1/programs/web.browser/program.json .meteor/local/build/programs/web.browser/program.json
5c5
<       "path": "4a77bd24e01b01378d1cd2099e98ed5ed328ff6f.js",
---
>       "path": "2de97dc80987b5a3279c0d84d23ebcc4b01b3e66.js",
9,11c9,11
<       "url": "/4a77bd24e01b01378d1cd2099e98ed5ed328ff6f.js",
<       "size": 1790630,
<       "hash": "4a77bd24e01b01378d1cd2099e98ed5ed328ff6f"
---
>       "url": "/2de97dc80987b5a3279c0d84d23ebcc4b01b3e66.js",
>       "size": 1790498,
>       "hash": "2de97dc80987b5a3279c0d84d23ebcc4b01b3e66"
```

Really obvious is that the non working version misses sourceMap and sourceMapRoot in .meteor/local/build/programs/server/program.json for meteorx.

**Comment by arunoda** [comment]
I've seen issues like these before. Frankly it's hard to tell and hard to debug. Let me know if you still getting these.

**Comment by peterkuiper** [comment]
Still getting these but I created a workaround by adding meteorx as a submodule under packages. It's not pretty but it is the only way this works. Note that this is only a problem when running meteor --production or running of a bundle.

If you have some insight on how to debug this that would be much appreciated!

**Comment by arunoda** [comment]
Try updating to the latest version of kadira. If not, give me a sample repo of this.
I try to work on this.

**Comment by peterkuiper** [comment]
We are already on the latest version of kadira:

```
cat .meteor/versions | grep meteorhacks:kadira
meteorhacks:kadira@2.23.6
```

The code is not public but perhaps I can replicate it with an empty project.

**Comment by peterkuiper** [comment]
Ok, I created a test repo:

```
git clone git@github.com:peterkuiper/meteorx-test.git
cd meteorx-test
meteor --production
```

You should see the following errors in the console:

Uncaught TypeError: Cannot read property 'MeteorX' of undefined
Uncaught TypeError: Cannot read property 'ServiceConfiguration' of undefined
Uncaught ReferenceError: Spacebars is not defined
Uncaught ReferenceError: Spacebars is not defined

To make the error go away:

```
mkdir packages
git clone git@github.com:meteorhacks/meteorx.git packages/meteorx
meteor --production
```

and no errors.

As I said, I'm not really into the internals but it seems like there is a loading order issue and maybe everything in packages has precendence over non-local packages. Do you know if this is true? I'd really like to know!

**Comment by lemonjp** [comment]
I got the same problem. +

**Comment by rfines** [comment]
I am also experiencing this issue with Meteor 1.1.0.3. It started happening when I updated all of the packages with meteor update --packages-only.

UPDATE:
Everything works fine locally doing what you mentioned above, but it doesn't fix the problem when I deploy the app to Heroku. The errors come right back after that. Is there something I need to do to deploy the apps with local packages?

FINAL UPDATE:
Apparently you cannot clone the package directly into the packages folder if you are going to be deploying to Heroku from git as the deploy tool will not traverse the link to the other repo. I had to download the zip file and copy/paste the appropriate files into the packages/meteorx directory in order to get the deployment to work on heroku.
